### PR TITLE
Changes to plan and allocation structs for preemption

### PR DESCRIPTION
**Pull request by preetapan**
This is a pretty small PR into the feature branch for internal struct and API level changes needed for preemption.

### CODE REVIEW

File: nomad/structs/structs.go [hunk 1]
```diff
@@ -645,6 +645,15 @@ type ApplyPlanResultsRequest struct {
 	// processed many times, potentially making state updates, without the state of
 	// the evaluation itself being updated.
 	EvalID string
+
+	// NodePreemptions is a map from node id to a set of allocations from other
```
Comment (dadgar) [inline nomad/structs/structs.go]: Comment is wrong, not actually a map

File: nomad/structs/structs.go [hunk 2]
```diff
@@ -1797,6 +1806,29 @@ func (r *Resources) Add(delta *Resources) error {
...
+	r.CPU -= delta.CPU
+	r.MemoryMB -= delta.MemoryMB
+	r.DiskMB -= delta.DiskMB
+	r.IOPS -= delta.IOPS
+
+	for _, n := range delta.Networks {
+		// Find the matching interface by IP or CIDR
+		idx := r.NetIndex(n)
+		if idx == -1 {
+			r.Networks = append(r.Networks, n.Copy())
```
Comment (dadgar) [inline nomad/structs/structs.go]: Can you put a comment on this? I don't understand it as is, if it doesn't exist, we add it? Seems like the opposite of subtract!

File: nomad/structs/structs.go [hunk 3]
```diff
@@ -7145,6 +7214,14 @@ func (p *Plan) IsNoOp() bool {
 		len(p.DeploymentUpdates) == 0
 }
 
+// PreemptedAllocs is used to store information about a set of allocations
+// for the same job that get preempted as part of placing allocations for the
+// job in the plan.
+
+// Preempted allocs represents a map from jobid to allocations
+// to be preempted
+type PreemptedAllocs map[*NamespacedID][]*Allocation
```
Comment (dadgar) [inline nomad/structs/structs.go]: Dead code?

File: nomad/structs/structs.go [hunk 4]
```diff
@@ -7195,6 +7277,8 @@ func (p *PlanResult) FullCommit(plan *Plan) (bool, int, int) {
 type PlanAnnotations struct {
 	// DesiredTGUpdates is the set of desired updates per task group.
 	DesiredTGUpdates map[string]*DesiredUpdates
+	// PreemptedAllocs is the set of allocations to be preempted to make the placement successful.
```
Comment (dadgar) [inline nomad/structs/structs.go]: Put a blank line between fields in structs when they have comments

File: nomad/structs/structs.go (hunk 2 above)
Comment (preetapan) [reply nomad/structs/structs.go]: This was a copy paste bug, it should only subtract if the resource exists for the way I am using this.

**Pull request merged**